File: src/burp_mcp_lite/filters.py
"""Filter helpers: status range parsing, host/path/method/mime predicates."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Callable, Iterable, Optional, Sequence, Union

from .snapshot import HistoryEntry


def parse_status_filter(spec: str) -> Callable[[int], bool]:
    """Parse a status filter spec.

    Accepted forms (comma-separated, AND-of-OR within a single arg):
      "200"            -> exact
      "2xx"            -> class
      "200,201,204"    -> set
      "400-499"        -> range (inclusive)
      "4xx-5xx"        -> two classes (200..599 covered)
      "200,4xx"        -> mix

    Empty / None handled by callers; this expects a non-empty string.
    """
    spec = spec.strip()
    matchers: list[Callable[[int], bool]] = []
    for token in spec.split(","):
        token = token.strip().lower()
        if not token:
            continue
        if "-" in token:
            lo_s, hi_s = token.split("-", 1)
            lo = _expand_low(lo_s)
            hi = _expand_high(hi_s)
            matchers.append(lambda s, lo=lo, hi=hi: lo <= s <= hi)
        elif token.endswith("xx") and len(token) == 3 and token[0].isdigit():
            cls = int(token[0])
            matchers.append(
                lambda s, cls=cls: (cls * 100) <= s <= (cls * 100 + 99)
            )
        else:
            try:
                exact = int(token)
            except ValueError:
                # unknown token: treat as never-match so the filter doesn't
                # silently let everything through.
                matchers.append(lambda s: False)
                continue
            matchers.append(lambda s, exact=exact: s == exact)
    if not matchers:
        return lambda s: True
    return lambda s: any(m(s) for m in matchers)


def _expand_low(tok: str) -> int:
    tok = tok.strip().lower()
    if tok.endswith("xx") and len(tok) == 3:
        return int(tok[0]) * 100
    return int(tok)


def _expand_high(tok: str) -> int:
    tok = tok.strip().lower()
    if tok.endswith("xx") and len(tok) == 3:
        return int(tok[0]) * 100 + 99
    return int(tok)


def _normalize_methods(m: Union[None, str, Sequence[str]]) -> Optional[set[str]]:
    if m is None:
        return None
    if isinstance(m, str):
        items = [x.strip() for x in m.split(",") if x.strip()]
    else:
        items = [x.strip() for x in m if x and x.strip()]
    if not items:
        return None
    return {x.upper() for x in items}


@dataclass
class HistoryFilter:
    host_substring: Optional[str] = None
    path_regex: Optional[re.Pattern] = None
    methods: Optional[set[str]] = None
    status_pred: Optional[Callable[[int], bool]] = None
    mime_substring: Optional[str] = None
    body_match_regex: Optional[re.Pattern] = None
    match_in: str = "response.body"

    def matches(self, e: HistoryEntry) -> bool:
        if self.methods and e.request.method.upper() not in self.methods:
            return False
        if self.host_substring and self.host_substring.lower() not in e.request.host.lower():
            return False
        if self.path_regex and not self.path_regex.search(e.request.path):
            return False
        if self.status_pred and not self.status_pred(e.response.status):
            return False
        if self.mime_substring:
            if self.mime_substring.lower() not in (e.response.content_type or "").lower():
                return False
        if self.body_match_regex:
            target = _resolve_match_target(e, self.match_in)
            if not self.body_match_regex.search(target):
                return False
        return True


def _resolve_match_target(e: HistoryEntry, target: str) -> str:
    target = target.strip().lower()
    if target == "response.body":
        return e.response.body
    if target == "response.headers":
        return _render_headers(e.response.headers)
    if target == "response.all":
        return _render_headers(e.response.headers) + "\n\n" + e.response.body
    if target == "request.body":
        return e.request.body
    if target == "request.headers":
        return _render_headers(e.request.headers)
    if target == "request.all":
        return _render_headers(e.request.headers) + "\n\n" + e.request.body
    return e.response.body


def _render_headers(headers: list[tuple[str, str]]) -> str:
    return "\n".join(f"{k}: {v}" for k, v in headers)


def build_filter(
    *,
    host: Optional[str] = None,
    path: Optional[str] = None,
    method: Union[None, str, Sequence[str]] = None,
    status: Optional[str] = None,
    mime: Optional[str] = None,
    match: Optional[str] = None,
    match_in: str = "response.body",
    case_sensitive: bool = False,
) -> HistoryFilter:
    flags = 0 if case_sensitive else re.IGNORECASE
    return HistoryFilter(
        host_substring=host,
        path_regex=re.compile(path, flags) if path else None,
        methods=_normalize_methods(method),
        status_pred=parse_status_filter(status) if status else None,
        mime_substring=mime,
        body_match_regex=re.compile(match, flags) if match else None,
        match_in=match_in,
    )


def apply_filter(entries: Iterable[HistoryEntry], f: HistoryFilter) -> list[HistoryEntry]:
    return [e for e in entries if f.matches(e)]

File: src/filters.ts
// Filter helpers: status range parsing, host/path/method/mime predicates.

import type { HistoryEntry } from "./snapshot.js";

export type StatusPredicate = (status: number) => boolean;

export type MatchTarget =
  | "request.body"
  | "request.headers"
  | "request.all"
  | "response.body"
  | "response.headers"
  | "response.all";

const STATUS_TOKEN_RX = /^([1-5])xx$/;

function expandLow(tok: string): number {
  const t = tok.trim().toLowerCase();
  const m = STATUS_TOKEN_RX.exec(t);
  if (m) return Number.parseInt(m[1]!, 10) * 100;
  return Number.parseInt(t, 10);
}

function expandHigh(tok: string): number {
  const t = tok.trim().toLowerCase();
  const m = STATUS_TOKEN_RX.exec(t);
  if (m) return Number.parseInt(m[1]!, 10) * 100 + 99;
  return Number.parseInt(t, 10);
}

export function parseStatusFilter(spec: string): StatusPredicate {
  const matchers: StatusPredicate[] = [];
  for (const rawTok of spec.split(",")) {
    const token = rawTok.trim().toLowerCase();
    if (!token) continue;

    if (token.includes("-")) {
      const [loS, hiS] = token.split("-", 2) as [string, string];
      const lo = expandLow(loS);
      const hi = expandHigh(hiS);
      if (Number.isFinite(lo) && Number.isFinite(hi)) {
        matchers.push((s) => s >= lo && s <= hi);
      } else {
        matchers.push(() => false);
      }
      continue;
    }

    const cls = STATUS_TOKEN_RX.exec(token);
    if (cls) {
      const c = Number.parseInt(cls[1]!, 10);
      matchers.push((s) => s >= c * 100 && s <= c * 100 + 99);
      continue;
    }

    const exact = Number.parseInt(token, 10);
    if (Number.isFinite(exact)) {
      matchers.push((s) => s === exact);
    } else {
      // Unknown token: treat as never-match so the filter doesn't silently
      // pass everything.
      matchers.push(() => false);
    }
  }
  if (matchers.length === 0) return () => true;
  return (s) => matchers.some((m) => m(s));
}

function normalizeMethods(
  m: string | readonly string[] | undefined,
): Set<string> | undefined {
  if (m === undefined) return undefined;
  const items =
    typeof m === "string"
      ? m.split(",").map((x) => x.trim()).filter(Boolean)
      : m.map((x) => x.trim()).filter(Boolean);
  if (items.length === 0) return undefined;
  return new Set(items.map((x) => x.toUpperCase()));
}

function renderHeaders(headers: readonly (readonly [string, string])[]): string {
  return headers.map(([k, v]) => `${k}: ${v}`).join("\n");
}

export function resolveMatchTarget(e: HistoryEntry, target: string): string {
  switch (target.trim().toLowerCase()) {
    case "response.body":
      return e.response.body;
    case "response.headers":
      return renderHeaders(e.response.headers);
    case "response.all":
      return renderHeaders(e.response.headers) + "\n\n" + e.response.body;
    case "request.body":
      return e.request.body;
    case "request.headers":
      return renderHeaders(e.request.headers);
    case "request.all":
      return renderHeaders(e.request.headers) + "\n\n" + e.request.body;
    default:
      return e.response.body;
  }
}

export interface BuildFilterArgs {
  host?: string;
  path?: string;
  method?: string | readonly string[];
  status?: string;
  mime?: string;
  match?: string;
  matchIn?: string;
  caseSensitive?: boolean;
}

export interface HistoryFilter {
  matches(e: HistoryEntry): boolean;
}

export function buildFilter(args: BuildFilterArgs = {}): HistoryFilter {
  const flags = args.caseSensitive ? "" : "i";
  const hostNeedle = args.host ? args.host.toLowerCase() : undefined;
  const pathRx = args.path ? new RegExp(args.path, flags) : undefined;
  const methods = normalizeMethods(args.method);
  const statusPred = args.status ? parseStatusFilter(args.status) : undefined;
  const mimeNeedle = args.mime ? args.mime.toLowerCase() : undefined;
  const matchRx = args.match ? new RegExp(args.match, flags) : undefined;
  const matchIn = args.matchIn ?? "response.body";

  return {
    matches(e) {
      if (methods && !methods.has((e.request.method || "").toUpperCase())) {
        return false;
      }
      if (hostNeedle) {
        if (!(e.request.host || "").toLowerCase().includes(hostNeedle)) {
          return false;
        }
      }
      if (pathRx && !pathRx.test(e.request.path || "")) return false;
      if (statusPred && !statusPred(e.response.status)) return false;
      if (mimeNeedle) {
        if (!(e.response.contentType || "").toLowerCase().includes(mimeNeedle)) {
          return false;
        }
      }
      if (matchRx) {
        const target = resolveMatchTarget(e, matchIn);
        if (!matchRx.test(target)) return false;
      }
      return true;
    },
  };
}

export function applyFilter(
  entries: readonly HistoryEntry[],
  f: HistoryFilter,
): HistoryEntry[] {
  return entries.filter((e) => f.matches(e));
}
